automated-comments/go/hamming/distance_signature_changed.md

```diff
@@ -0,0 +1,4 @@
+The signature of the `Distance` function should be `Distance(a, b string) (int, error)`.
+Exercism is built on the TDD (Test Driven Development) principle.
```

Comment:
This is absolutely right, but a key part of TDD is also updating the tests to take account of changes to the code under test!

A simple example is in `grains`, when students often implement Total() by calling Square repeatedly and ignoring the error value returned. I sometimes guide them towards checking the error and returning it if non-nil, which of course changes the function signature and thus requires updating the tests.

I'd be cautious about discouraging students from changing the function signature _when they have a good reason_ for doing so. Perhaps say rather "The test-driven development (TDD) method which Exercism uses means that the function signature must always match what the tests expect. If the function signature is updated, the tests which call it must also be updated to match, or they won't work."

Reply:
@bitfield not to disregard your comment, which I think is valid, I also agree very much with @tehsphinx for the simple reason that we have one more constraint: we need to be able to make some assumptions about the code when analyzing it, for now. 

Discouraging changing tests is one way to be able to do that 😅.

Is it common here to change the tests? If it's not common and not _expected_, I would discourage it, like Gabriel currently does.

Reply:
Generally I think this is a valid point but as @SleeplessByte said: We are not really able to incorporate that into the analysers.

For now this comment will only be visible to mentors anyway as it is a `disapprove` comment. It is up to the mentor to decide if he ignores it and works with the student on a changed function signature or not.

I'd also argue, that the function signatures in the tests are pretty well thought through and if a change is regularly required we should change that in the exercises.

Reply:
Personally, I usually discourage changing the tests, as it makes communicating with the student harder for the mentor, which can no longer assume he knows how the tests are structured. It also makes it harder for the analyzer to work :)